pkg/arvo/app/chat-store.hoon

```diff
@@ -0,0 +1,238 @@
+::
+/+  *chat-json
+|%
++$  move  [bone card]
+::
++$  card
+  $%  [%diff diff]
+      [%quit ~]
+  ==
+::
++$  state
+  $%  [%0 state-zero]
+  ==
+::
++$  state-zero
+  $:  =inbox
+  ==
+::
++$  diff
+  $%  [%chat-initial inbox]
+      [%chat-configs chat-configs]
+      [%chat-update chat-update]
+  ==
+--
+::
+|_  [bol=bowl:gall state]
+::
+++  this  .
+::
+++  prep
+  |=  old=(unit state)
+  ^-  (quip move _this)
+  [~ ?~(old this this(+<+ u.old))]
+::
+++  peek-x-all
+  |=  pax=path
+  ^-  (unit (unit [%noun (map path mailbox)]))
+  [~ ~ %noun inbox]
+::
+++  peek-x-configs
+  |=  pax=path
+  ^-  (unit (unit [%noun chat-configs]))
+  :^  ~  ~  %noun
+  (inbox-to-configs inbox)
+::
+++  peek-x-keys
+  |=  pax=path
+  ^-  (unit (unit [%noun (set path)]))
+  [~ ~ %noun ~(key by inbox)]
+::
+++  peek-x-mailbox
+  |=  pax=path
+  ^-  (unit (unit [%noun (unit mailbox)]))
+  ?~  pax  ~
+  =/  mailbox=(unit mailbox)  (~(get by inbox) pax)
+  [~ ~ %noun mailbox]
+::
+++  peek-x-config
+  |=  pax=path
+  ^-  (unit (unit [%noun config]))
+  ?~  pax  ~
+  =/  mailbox  (~(get by inbox) pax)
+  ?~  mailbox  ~
+  :^  ~  ~  %noun
+  config.u.mailbox
+::
+++  peek-x-envelopes
+  |=  pax=path
+  ^-  (unit (unit [%noun (list envelope)]))
+  ?+  pax  ~
+      [@ @ *]
+    =/  mail-path  t.t.pax
+    =/  mailbox  (~(get by inbox) mail-path)
+    ?~  mailbox
+      [~ ~ %noun ~]
+    =*  envelopes  envelopes.u.mailbox
+    =/  sign-test=[?(%neg %pos) @]
+      %-  need
+      %+  rush  i.pax
+      ;~  pose
+        %+  cook
+          |=  n=@
+          [%neg n]
+        ;~(pfix hep dem:ag)
+      ::
+        %+  cook
+          |=  n=@
+          [%pos n]
+        dem:ag
+      ==
+    =*  length  length.config.u.mailbox
+    =*  start  +.sign-test
+    ?:  =(-.sign-test %neg)
+      ?:  (gth start length)
+        [~ ~ %noun envelopes]
+      [~ ~ %noun (swag [(sub length start) start] envelopes)]
+    ::
+    =/  end  (slav %ud i.t.pax)
+    ?.  (lte start end)
+      ~
+    =.  end  ?:((lth end length) end length)
+    [~ ~ %noun (swag [start (sub end start)] envelopes)]
+  ==
+::
+++  peer-keys
+  |=  pax=path
+  ^-  (quip move _this)
+  ?>  (team:title our.bol src.bol)
+  ::  we send the list of keys then send events when they change
+  :_  this
+  [ost.bol %diff %chat-update [%keys ~(key by inbox)]]~
+::
+++  peer-all
+  |=  pax=path
+  ^-  (quip move _this)
+  ?>  (team:title our.bol src.bol)
+  :_  this
+  [ost.bol %diff %chat-initial inbox]~
+::
+++  peer-configs
+  |=  pax=path
+  ^-  (quip move _this)
+  ?>  (team:title our.bol src.bol)
+  :_  this
+  [ost.bol %diff %chat-configs (inbox-to-configs inbox)]~
+::
+++  peer-updates
+  |=  pax=path
+  ^-  (quip move _this)
+  ?>  (team:title our.bol src.bol)
+  ::  we now proxy all events to this path
+  [~ this]
+::
+++  peer-mailbox
+  |=  pax=path
+  ^-  (quip move _this)
+  ?>  (team:title our.bol src.bol)
+  =/  box=(unit mailbox)  (~(get by inbox) pax)
+  ?~  box  !!
```

Comment:
got:by does this

Reply:
Modified here and in other places. Thanks.